File: llm/llama_cpp_unified.go
package llm

import (
	"bufio"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/ollama/ollama/internal/reporoots"
)

const (
	defaultUnifiedLlamaCppCommit = "c84b30200c8d512c00c9d61c96bed078f1c0024d"
	unifiedLlamaCppRepo          = "https://github.com/elizaOS/llama.cpp.git"
)

// UnifiedLlamaCppRoot is the single runtime llama.cpp checkout.
//
// Order: LLAMA_CPP_ROOT → vendor/llama-cpp-<pin> (patched) → ../llama.cpp sibling.
// WHY vendor default: build_llama_server.sh and in-process ggml share Ollama patches;
// bare eliza sibling alone misses kv-ext / compat hooks.
func UnifiedLlamaCppRoot() string {
	if raw := strings.TrimSpace(os.Getenv("LLAMA_CPP_ROOT")); raw != "" {
		return filepath.Clean(raw)
	}
	if vendor := vendorLlamaCppRoot(); vendor != "" {
		return vendor
	}
	if repo := zerollamaRepoRoot(); repo != "" {
		return filepath.Clean(filepath.Join(repo, "..", "llama.cpp"))
	}
	return ""
}

func vendorLlamaCppRoot() string {
	repo := zerollamaRepoRoot()
	if repo == "" {
		return ""
	}
	fetchHead := readVendorLlamaCppPin()
	if fetchHead == "" {
		return ""
	}
	vendor := filepath.Join(repo, "vendor", "llama-cpp-"+fetchHead)
	if _, err := os.Stat(filepath.Join(vendor, "CMakeLists.txt")); err != nil {
		return ""
	}
	return vendor
}

func zerollamaRepoRoot() string {
	for _, root := range reporoots.SearchRoots() {
		if _, err := os.Stat(filepath.Join(root, "go.mod")); err != nil {
			continue
		}
		if _, err := os.Stat(filepath.Join(root, "LLAMA_CPP_COMMIT")); err == nil {
			return root
		}
	}
	return ""
}

// PinnedLlamaCppCommit reads repo-root LLAMA_CPP_COMMIT (unified runtime pin).
func PinnedLlamaCppCommit() string {
	repo := zerollamaRepoRoot()
	if repo == "" {
		return defaultUnifiedLlamaCppCommit
	}
	path := filepath.Join(repo, "LLAMA_CPP_COMMIT")
	data, err := os.ReadFile(path)
	if err != nil {
		return defaultUnifiedLlamaCppCommit
	}
	ref := strings.TrimSpace(string(data))
	if ref == "" {
		return defaultUnifiedLlamaCppCommit
	}
	return ref
}

func siblingLlamaCppRoot() string {
	repo := zerollamaRepoRoot()
	if repo == "" {
		return ""
	}
	return filepath.Clean(filepath.Join(repo, "..", "llama.cpp"))
}

func isBareSiblingLlamaRoot(path string) bool {
	if path == "" {
		return false
	}
	sibling := siblingLlamaCppRoot()
	return sibling != "" && filepath.Clean(path) == sibling
}

func llamaServerBinUnderRoot(bin, root string) bool {
	if bin == "" || root == "" {
		return false
	}
	rel, err := filepath.Rel(filepath.Clean(root), filepath.Clean(bin))
	return err == nil && !strings.HasPrefix(rel, "..")
}

func unifiedLlamaServerBinForRoot(root string) (string, bool) {
	if root == "" {
		return "", false
	}
	path := filepath.Join(root, "build", "bin", llamaCppBinaryName("llama-server", runtime.GOOS))
	if st, err := os.Stat(path); err == nil && !st.IsDir() {
		return path, true
	}
	return path, false
}

// UnifiedLlamaServerBin returns $LLAMA_CPP_ROOT/build/bin/llama-server when present.
func UnifiedLlamaServerBin() string {
	root := UnifiedLlamaCppRoot()
	if root == "" {
		return ""
	}
	path, _ := unifiedLlamaServerBinForRoot(root)
	return path
}

func unifiedLlamaServerBinExists() (string, bool) {
	if vendor := vendorLlamaCppRoot(); vendor != "" {
		if path, ok := unifiedLlamaServerBinForRoot(vendor); ok {
			return path, true
		}
	}
	return unifiedLlamaServerBinForRoot(UnifiedLlamaCppRoot())
}

// IsLegacyLlamaCppCheckout reports deprecated sibling names (eliza-llama.cpp, etc.).
func IsLegacyLlamaCppCheckout(path string) bool {
	if path == "" {
		return false
	}
	switch filepath.Base(filepath.Clean(path)) {
	case "eliza-llama.cpp", "eliza_llama.cpp", "stock-llama.cpp", "ollama-upstream":
		return true
	default:
		return false
	}
}

// LlamaCppUnificationStatus summarizes pin + binary layout for doctor / status.
type LlamaCppUnificationStatus struct {
	UnifiedRoot       string
	PinnedCommit      string
	SiblingHEAD       string
	EffectiveServer   string
	ServerFromEnv     bool
	ServerUnderRoot   bool
	LegacyCheckout    bool
	VendorPin         string
	RuntimeReady      bool
	Detail            string
	Warn              bool
	FixHint           string
}

func LlamaCppUnificationReport() LlamaCppUnificationStatus {
	root := UnifiedLlamaCppRoot()
	pin := PinnedLlamaCppCommit()
	report := LlamaCppUnificationStatus{
		UnifiedRoot:  root,
		PinnedCommit: pin,
		VendorPin:    readVendorLlamaCppPin(),
	}

	if root != "" {
		report.LegacyCheckout = IsLegacyLlamaCppCheckout(root)
		report.SiblingHEAD = gitRevParseHEAD(root)
		if path, ok := unifiedLlamaServerBinExists(); ok {
			report.RuntimeReady = true
			report.EffectiveServer = path
		}
	}

	if override := strings.TrimSpace(os.Getenv("LLAMA_SERVER_BIN")); override != "" {
		report.ServerFromEnv = true
		report.EffectiveServer = filepath.Clean(override)
	}

	if report.EffectiveServer != "" && root != "" {
		rel, err := filepath.Rel(root, report.EffectiveServer)
		report.ServerUnderRoot = err == nil && !strings.HasPrefix(rel, "..")
	}

	var parts []string
	parts = append(parts, "unified "+root+" @ pin "+shortRef(pin))
	if report.SiblingHEAD != "" {
		parts = append(parts, "HEAD "+shortRef(report.SiblingHEAD))
	}
	if report.EffectiveServer != "" {
		parts = append(parts, "llama-server "+report.EffectiveServer)
	}
	if report.VendorPin != "" && !pinRefsMatch(report.VendorPin, pin) {
		parts = append(parts, "in-process vendor "+shortRef(report.VendorPin)+" (rebase pending)")
		report.Warn = true
	}
	if report.LegacyCheckout {
		parts = append(parts, "legacy checkout name — migrate to vendor tree")
		report.Warn = true
		report.FixHint = "./scripts/rebase_vendor_unified.sh --sync && ./scripts/build_llama_server.sh"
	}
	if vendor := vendorLlamaCppRoot(); vendor != "" && isBareSiblingLlamaRoot(root) {
		parts = append(parts, "LLAMA_CPP_ROOT bare sibling — prefer "+vendor)
		report.Warn = true
		if report.FixHint == "" {
			report.FixHint = "unset LLAMA_CPP_ROOT or export LLAMA_CPP_ROOT=" + vendor
		}
	}
	if report.ServerFromEnv && report.EffectiveServer != "" {
		if vendor := vendorLlamaCppRoot(); vendor != "" && !llamaServerBinUnderRoot(report.EffectiveServer, vendor) {
			if vPath, ok := unifiedLlamaServerBinForRoot(vendor); ok {
				parts = append(parts, "patched vendor llama-server available at "+vPath)
				report.Warn = true
				if report.FixHint == "" {
					report.FixHint = "unset LLAMA_SERVER_BIN (zerollama prefers vendor build)"
				}
			}
		}
	}
	if report.ServerFromEnv && !report.ServerUnderRoot && report.EffectiveServer != "" && !report.Warn {
		parts = append(parts, "LLAMA_SERVER_BIN outside unified tree")
		report.Warn = true
		if report.FixHint == "" {
			report.FixHint = "unset LLAMA_SERVER_BIN or point at " + filepath.Join(root, "build", "bin", "llama-server")
		}
	}
	if !report.RuntimeReady && root != "" {
		parts = append(parts, "llama-server not built")
		report.Warn = true
		if report.FixHint == "" {
			report.FixHint = "./scripts/build_llama_server.sh"
		}
	}
	report.Detail = strings.Join(parts, "; ")
	return report
}

func readVendorLlamaCppPin() string {
	repo := zerollamaRepoRoot()
	if repo == "" {
		return ""
	}
	if data, err := os.ReadFile(filepath.Join(repo, "LLAMA_CPP_VERSION")); err == nil {
		if v := strings.TrimSpace(string(data)); v != "" {
			return v
		}
	}
	f, err := os.Open(filepath.Join(repo, "Makefile.sync"))
	if err != nil {
		return ""
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if strings.HasPrefix(line, "FETCH_HEAD=") {
			return strings.TrimPrefix(line, "FETCH_HEAD=")
		}
	}
	return ""
}

func pinRefsMatch(shortPin, fullPin string) bool {
	shortPin = strings.TrimSpace(shortPin)
	fullPin = strings.TrimSpace(fullPin)
	if shortPin == "" || fullPin == "" {
		return false
	}
	if shortPin == fullPin {
		return true
	}
	return strings.HasPrefix(fullPin, shortPin)
}

func LlamaCppPathUsesLegacyCheckout(path string) bool {
	if path == "" {
		return false
	}
	clean := filepath.Clean(path)
	for _, part := range strings.Split(clean, string(os.PathSeparator)) {
		if IsLegacyLlamaCppCheckout(part) {
			return true
		}
	}
	return false
}

// ApplyUnifiedLlamaCppEnv redirects legacy LLAMA_CPP_ROOT / LLAMA_SERVER_BIN to the
// patched vendor tree when available.
//
// WHY: operators often keep shell exports pointing at ../eliza-llama.cpp or bare
// ../llama.cpp after U3 — MTP/eagle3 then miss Ollama patches and argv probes fail.
func ApplyUnifiedLlamaCppEnv() []string {
	var msgs []string
	vendor := vendorLlamaCppRoot()
	unified := UnifiedLlamaCppRoot()

	if vendor != "" {
		root := strings.TrimSpace(os.Getenv("LLAMA_CPP_ROOT"))
		switch {
		case root == "":
			_ = os.Setenv("LLAMA_CPP_ROOT", vendor)
			unified = vendor
		case LlamaCppPathUsesLegacyCheckout(root) || isBareSiblingLlamaRoot(root):
			_ = os.Setenv("LLAMA_CPP_ROOT", vendor)
			msgs = append(msgs, "set LLAMA_CPP_ROOT="+vendor+" (was "+root+")")
			unified = vendor
		}
	}

	serverPath, serverOK := unifiedLlamaServerBinForRoot(unified)
	if !serverOK && vendor != "" && unified != vendor {
		if path, ok := unifiedLlamaServerBinForRoot(vendor); ok {
			serverPath, serverOK = path, true
		}
	}

	if root := strings.TrimSpace(os.Getenv("LLAMA_CPP_ROOT")); root != "" && LlamaCppPathUsesLegacyCheckout(root) && vendor == "" {
		msgs = append(msgs, "LLAMA_CPP_ROOT uses legacy checkout "+root)
		if unified != "" {
			if _, err := os.Stat(filepath.Join(unified, "CMakeLists.txt")); err == nil {
				_ = os.Setenv("LLAMA_CPP_ROOT", unified)
				msgs = append(msgs, "set LLAMA_CPP_ROOT="+unified)
			}
		}
	}

	if bin := strings.TrimSpace(os.Getenv("LLAMA_SERVER_BIN")); bin != "" {
		redirect := LlamaCppPathUsesLegacyCheckout(bin)
		if !redirect && vendor != "" && serverOK && !llamaServerBinUnderRoot(bin, vendor) {
			redirect = isBareSiblingLlamaRoot(rootPathForLlamaServerBin(bin)) || LlamaCppPathUsesLegacyCheckout(bin)
		}
		if redirect && serverOK {
			msgs = append(msgs, "LLAMA_SERVER_BIN was "+bin)
			_ = os.Setenv("LLAMA_SERVER_BIN", serverPath)
			msgs = append(msgs, "set LLAMA_SERVER_BIN="+serverPath)
		} else if LlamaCppPathUsesLegacyCheckout(bin) && !serverOK && unified != "" {
			msgs = append(msgs, "run ./scripts/build_llama_server.sh in "+unified)
		}
	}

	if strings.TrimSpace(os.Getenv("LLAMA_CPP_ROOT")) == "" && unified != "" {
		if _, err := os.Stat(filepath.Join(unified, "CMakeLists.txt")); err == nil {
			_ = os.Setenv("LLAMA_CPP_ROOT", unified)
		}
	}
	if strings.TrimSpace(os.Getenv("LLAMA_CPP_LIB")) == "" && unified != "" {
		libName := "libllama.so"
		if runtime.GOOS == "darwin" {
			libName = "libllama.dylib"
		} else if runtime.GOOS == "windows" {
			libName = "llama.dll"
		}
		lib := filepath.Join(unified, "build", "bin", libName)
		if st, err := os.Stat(lib); err == nil && !st.IsDir() {
			_ = os.Setenv("LLAMA_CPP_LIB", lib)
		}
	}
	if strings.TrimSpace(os.Getenv("LLAMA_SERVER_BIN")) == "" && serverOK {
		_ = os.Setenv("LLAMA_SERVER_BIN", serverPath)
	}

	return msgs
}

func rootPathForLlamaServerBin(bin string) string {
	// .../build/bin/llama-server → checkout root
	return filepath.Clean(filepath.Join(filepath.Dir(bin), "..", ".."))
}

func gitRevParseHEAD(root string) string {
	cmd := exec.Command("git", "-C", root, "rev-parse", "HEAD")
	out, err := cmd.Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

func shortRef(ref string) string {
	ref = strings.TrimSpace(ref)
	if len(ref) > 12 {
		return ref[:12]
	}
	return ref
}

// UnifiedLlamaCppRepo is the canonical clone URL for ensure_llama_cpp_sibling.sh parity.
func UnifiedLlamaCppRepo() string {
	if raw := strings.TrimSpace(os.Getenv("LLAMA_CPP_REPO")); raw != "" {
		return raw
	}
	return unifiedLlamaCppRepo
}

File: runtime/runtime/llama_cpp_unified.py
"""Unified llama.cpp root resolution (parity with llm/llama_cpp_unified.go).

WHY: runtime subprocess + in-process must agree with Go serve on which checkout
is canonical — one ../llama.cpp @ LLAMA_CPP_COMMIT, not legacy eliza-llama siblings.
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Any

from runtime.llama_cpp_probe import default_llama_cpp_root, pinned_llama_cpp_version

LEGACY_CHECKOUT_NAMES = frozenset(
    {
        "eliza-llama.cpp",
        "eliza_llama.cpp",
        "stock-llama.cpp",
        "ollama-upstream",
    }
)

DEFAULT_PIN = "8f114a9b573b69035299f9b924047f53c1e22c7e"
UNIFIED_REPO = "https://github.com/ggml-org/llama.cpp.git"


def _repo_root() -> Path:
    return Path(__file__).resolve().parents[2]


def _sibling_llama_cpp_root() -> Path:
    return (_repo_root().parent / "llama.cpp").resolve()


def _is_bare_sibling_root(path: str | Path) -> bool:
    try:
        return Path(path).resolve() == _sibling_llama_cpp_root()
    except OSError:
        return False


def _server_under_root(bin_path: str | Path, root: str) -> bool:
    try:
        Path(bin_path).resolve().relative_to(Path(root).resolve())
        return True
    except ValueError:
        return False


def is_legacy_checkout(path: str | Path | None) -> bool:
    if not path:
        return False
    return Path(path).name in LEGACY_CHECKOUT_NAMES


def path_uses_legacy_checkout(path: str | Path | None) -> bool:
    if not path:
        return False
    return any(is_legacy_checkout(part) for part in Path(path).parts)


def unified_llama_server_bin(root: Path | None = None) -> Path | None:
    base = root or default_llama_cpp_root()
    if not base.is_dir():
        return None
    for name in ("llama-server", "llama-server.exe"):
        candidate = base / "build" / "bin" / name
        if candidate.is_file():
            return candidate
    return None


def vendor_llama_cpp_root(repo_root: Path | None = None) -> Path | None:
    """Pinned vendor tree under ``zerollama/vendor/llama-cpp-<pin>``."""
    from runtime.llama_cpp_probe import _vendor_llama_cpp_root

    return _vendor_llama_cpp_root(repo_root)


def resolve_llama_cpp_root() -> Path:
    """Canonical llama.cpp root for runtime + subprocess.

    WHY prefer vendor over bare sibling ``../llama.cpp``: zerollama patches
    (``POST /kv/seq-copy``, cuda-graph invalidate) ship on the vendor pin only;
    stale ``LLAMA_CPP_ROOT`` in the shell must not override the patched tree.
    """
    return default_llama_cpp_root()


def _unique_roots(*roots: Path | None) -> list[Path]:
    out: list[Path] = []
    seen: set[str] = set()
    for r in roots:
        if r is None:
            continue
        try:
            key = str(r.resolve())
        except OSError:
            continue
        if key in seen:
            continue
        seen.add(key)
        out.append(Path(key))
    return out


def resolve_llama_server_bin(root: Path | None = None) -> Path | None:
    """``llama-server`` path: explicit env → primary root → vendor fallback."""
    override = os.environ.get("LLAMA_SERVER_BIN", "").strip()
    if override:
        p = Path(override).expanduser()
        if not p.is_absolute():
            p = p.resolve()
        return p if p.is_file() else None
    primary = root or resolve_llama_cpp_root()
    for r in _unique_roots(primary, vendor_llama_cpp_root()):
        server = unified_llama_server_bin(r)
        if server is not None:
            return server
    return None


def resolve_llama_cpp_lib(root: Path | None = None) -> Path | None:
    """``libllama`` for in-process backend; follows same root order as server bin."""
    override = os.environ.get("LLAMA_CPP_LIB", "").strip()
    if override:
        p = Path(override).expanduser()
        if not p.is_absolute():
            p = p.resolve()
        return p if p.is_file() else None
    primary = root or resolve_llama_cpp_root()
    names = ("libllama.so", "libllama.dylib", "libllama.dll")
    for r in _unique_roots(primary, vendor_llama_cpp_root()):
        for name in names:
            candidate = r / "build" / "bin" / name
            if candidate.is_file():
                return candidate
    return None


def unified_health(
    *,
    llama_cpp_root: Path | str | None = None,
    llama_server_bin: Path | str | None = None,
) -> dict[str, Any]:
    """Health block for /health.llama_cpp_unified."""
    root = Path(llama_cpp_root) if llama_cpp_root else default_llama_cpp_root()
    root_s = str(root)
    pin = pinned_llama_cpp_version() or DEFAULT_PIN
    server = Path(llama_server_bin) if llama_server_bin else unified_llama_server_bin(root)
    server_s = str(server) if server else None

    legacy_root = is_legacy_checkout(root) or path_uses_legacy_checkout(root_s)
    legacy_server = path_uses_legacy_checkout(server_s)
    warn = legacy_root or legacy_server

    detail_parts = [f"unified {root_s} @ pin {pin[:12]}"]
    if server_s:
        detail_parts.append(f"llama-server {server_s}")
    if legacy_root:
        detail_parts.append("legacy LLAMA_CPP_ROOT name — migrate to ../llama.cpp")
        warn = True
    if legacy_server:
        detail_parts.append("LLAMA_SERVER_BIN outside unified tree or legacy name")
        warn = True
    if server is None:
        detail_parts.append("llama-server not built")
        warn = True

    return {
        "unified_root": root_s,
        "pinned_commit": pin,
        "llama_server_bin": server_s,
        "legacy_checkout": legacy_root or legacy_server,
        "runtime_ready": server is not None and server.is_file(),
        "repo": UNIFIED_REPO,
        "warn": warn,
        "detail": "; ".join(detail_parts),
        "migrate": "./scripts/migrate_llama_cpp_unified.sh",
    }


def normalize_llama_cpp_env() -> list[str]:
    """Return human-readable normalization notes (does not mutate os.environ)."""
    msgs: list[str] = []
    root = resolve_llama_cpp_root()
    env_root = os.environ.get("LLAMA_CPP_ROOT", "").strip()
    env_bin = os.environ.get("LLAMA_SERVER_BIN", "").strip()
    server = unified_llama_server_bin(root)

    if env_root and path_uses_legacy_checkout(env_root):
        msgs.append(f"LLAMA_CPP_ROOT legacy {env_root} → use {root}")
    if env_root and _is_bare_sibling_root(env_root):
        msgs.append(f"LLAMA_CPP_ROOT bare sibling {env_root} → use {root}")
    if env_bin and path_uses_legacy_checkout(env_bin):
        if server:
            msgs.append(f"LLAMA_SERVER_BIN legacy {env_bin} → use {server}")
        else:
            msgs.append(f"LLAMA_SERVER_BIN legacy; build {root}/build/bin/llama-server")
    elif env_bin and server and not _server_under_root(env_bin, str(root)):
        if root.name.startswith("llama-cpp-"):
            msgs.append(f"LLAMA_SERVER_BIN {env_bin} → prefer {server}")
    return msgs
